### PR TITLE
Bugfix/blockstore stuck issue 317

### LINKED ISSUE #317: BlockStore stuck - need logs
There is very rare bug in BlockStore during IBD. I've seen it several times before. I've tried to run about 100 IBDs today to try to confirm it before the sprint planning and I've failed, it did not occur once. But now testing the outcome of a new pull request, it is there (again).

There is not much we can (reasonably) do about it without having full logs in BlockStore. Maybe alternative is to review the code and search for it like that, but you can easily miss it that way.

I suggest we just log it.

What does it do? It simply creates a big gap during IBD between Consensus and Store:

```
      ======Node stats====== 08/14/2017 15:50:46 agent StratisBitcoin:1.0.2
      Headers.Height:    480506   Headers.Hash:     000000000000000000807de0ce829b6eb4dc04545d9768fc84088c2eb3d9940d
      Consensus.Height:  114071   Consensus.Hash:   000000000000478f1413014c8a6d85652be5103b74d75eb1adce38974922c89a
      Store.Height:      55811    Store.Hash:       0000000013d53405392513d5e7d26630c0231ace81d404224f90dedba9c6e7dd
```

Once the Store is stuck, it will stay there. I thought previously restarting the node helped, but currently I can see it does not. Which is good - I've got instance of that and if it stays like this it should be easy to find.

### PR DESCRIPTION
**Pull request by Aprogiena**
Introduction of 20 second flush time for DownloadStack together with increasing the size limit from 5 to 20 MB, together with reducing max stall count from 10k to 1.8k

Fix #317

### CODE REVIEW

**Comment by Aprogiena** [comment]
tests are in progress as we speak

File: Stratis.Bitcoin/Features/BlockStore/BlockStoreLoop.cs [hunk 1]
```diff
@@ -18,6 +18,12 @@ namespace Stratis.Bitcoin.Features.BlockStore
     /// </summary>
     public class BlockStoreLoop
     {
+        /// <summary>Maximal number of bytes the block puller can download before the downloaded blocks are stored to the disk.</summary>
```
Comment (fassadlr) [inline Stratis.Bitcoin/Features/BlockStore/BlockStoreLoop.cs]: This should be Maximum instead if Maximal

File: Stratis.Bitcoin/Features/BlockStore/BlockStoreLoop.cs [hunk 2]
```diff
@@ -18,6 +18,12 @@ namespace Stratis.Bitcoin.Features.BlockStore
...
+        internal const uint MaxInsertBlockSize = 20 * 1024 * 1024;
+
+        /// <summary>Maximal number of bytes the pending storage can hold until the downloaded blocks are stored to the disk.</summary>
```
Comment (fassadlr) [inline Stratis.Bitcoin/Features/BlockStore/BlockStoreLoop.cs]: This should be Maximum instead if Maximal

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreInnerStepContext.cs [hunk 3]
```diff
@@ -16,18 +18,35 @@ namespace Stratis.Bitcoin.Features.BlockStore.LoopSteps
...
+
         /// <summary>Instance logger.</summary>
         private readonly ILogger logger;
 
         /// <summary>Factory for creating loggers.</summary>
         private readonly ILoggerFactory loggerFactory;
 
-        public BlockStoreInnerStepContext(CancellationToken cancellationToken, BlockStoreLoop blockStoreLoop, ILoggerFactory loggerFactory)
+        /// <summary>Provider of time functions.</summary>
+        internal readonly IDateTimeProvider DateTimeProvider;
```
Comment (fassadlr) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreInnerStepContext.cs]: Make this a property with a private setter:

`internal readonly IDateTimeProvider DateTimeProvider { get; private set;}`

File: Stratis.Bitcoin/Features/BlockStore/BlockStoreLoop.cs (hunk 1 above)
Comment (Aprogiena) [inline Stratis.Bitcoin/Features/BlockStore/BlockStoreLoop.cs]: yes, tx

File: Stratis.Bitcoin/Features/BlockStore/BlockStoreLoop.cs (hunk 2 above)
Comment (Aprogiena) [inline Stratis.Bitcoin/Features/BlockStore/BlockStoreLoop.cs]: yes tx

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreInnerStepContext.cs (hunk 3 above)
Comment (Aprogiena) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreInnerStepContext.cs]: yes, will fix

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreInnerStepDownloadBlocks.cs [hunk 4]
```diff
@@ -52,14 +53,22 @@ public BlockStoreInnerStepDownloadBlocks(ILoggerFactory loggerFactory)
                 context.InsertBlockSize += downloadedBlock.Length;
                 context.StallCount = 0;
 
-                this.logger.LogTrace("Insert block size is {0} bytes, download stack contains {1} more blocks to download.", context.InsertBlockSize, context.DownloadStack.Count);
-                if (context.InsertBlockSize > context.BlockStoreLoop.InsertBlockSizeThreshold || !context.DownloadStack.Any())
+                DateTime now = context.DateTimeProvider.GetUtcNow();
+                uint lastFlushDiff = (uint)(now - context.LastDownloadStackFlushTime).TotalMilliseconds;
+                this.logger.LogTrace("Insert block size is {0} bytes, download stack contains {1} more blocks to download, last flush time was {2} ms ago.", context.InsertBlockSize, context.DownloadStack.Count, lastFlushDiff);
+
```
Comment (dangershony) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreInnerStepDownloadBlocks.cs]: I love it, its just so much more readable.

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/ProcessPendingStorageStep.cs [hunk 5]
```diff
@@ -88,9 +88,10 @@ internal ProcessPendingStorageStep(BlockStoreLoop blockStoreLoop, ILoggerFactory
                     lastFoundChainedBlock = nextChainedBlock;
                 }
 
-                if (pendingStorageBatchSize > this.BlockStoreLoop.InsertBlockSizeThreshold || breakExecution)
+                // TODO: breakExecution here is always false.
```
Comment (dangershony) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/ProcessPendingStorageStep.cs]: I am not sure this is true.
> TODO: breakExecution here is always false.

Comment (Aprogiena) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/ProcessPendingStorageStep.cs]: you are right, not sure what i was smoking, this comment line should be removed

**Pull request merged**